dbs-boot: arm add dbs-boot crate

jingshan-lx opened the pull request:
This PR adds the arm dbs-boot crate, which mainly involves three modifications:
1. Add arm layout.rs.
2. Add fdt.rs to inform the guest kernel device information.
3. Add the dtb file for unit-test to ensure that the fdt build is basically correct.

jingshan-lx committed:
dbs-boot: arm add layout.rs

Align the x86 architecture, put arm layout.rs in the dbs-boot directory,
and dbs-boot depends on dbs-arch.

Since the gic module is stored in the dbs-arch directory, and the
virtualized gic needs to specify the mmio address space of the gic.
Therefore, the mmio address space of gic is first defined in the gic
module, and on this basis, other contents in layout.rs are continued to
be constructed.

In addition, we define IRQ_BASE and IRQ_MAX also depend on the relevant
definitions in gic.

Signed-off-by: jingshan.lx <jingshan@linux.alibaba.com>

jingshan-lx committed:
dbs-boot: arm add fdt feature

Arm adds the fdt feature, which is mainly used to inform the guest
kernel of device information.

Signed-off-by: jingshan.lx <jingshan@linux.alibaba.com>

jingshan-lx committed:
dbs-boot: arm add dtb files

Arm adds dtb files. These two files are mainly used for unit-test to
ensure that the basic functions of fdt are correct.

Signed-off-by: jingshan.lx <jingshan@linux.alibaba.com>

jingshan-lx committed:
dbs-boot: arm add fdt dependency library

This modification mainly adds arm fdt dependency library and exposes fdt to
external modules.

Signed-off-by: jingshan.lx <jingshan@linux.alibaba.com>

jiangliu commented on crates/dbs-boot/src/lib.rs:
```diff
@@ -7,21 +7,40 @@
 
+/// Types of devices that can get attached to this platform.
+#[derive(Clone, Debug, PartialEq, Eq, Hash, Copy)]
+pub enum DeviceType {
```
Seem it would be better to move this into dbs-arch?

jingshan-lx replied:
done

jiangliu commented on crates/dbs-boot/src/aarch64/fdt.rs:
```diff
@@ -0,0 +1,774 @@
+}
+
+/// Trait for devices to be added to the Flattened Device Tree.
+pub trait DeviceInfoForFDT {
```
Should this be moved into dbs-arch?

jingshan-lx replied:
done

jiangliu commented on crates/dbs-boot/src/aarch64/fdt.rs:
```diff
@@ -0,0 +1,774 @@
+
+/// MMIO device info used for FDT generating.
+#[derive(Debug, Clone, PartialEq)]
+pub struct MMIODeviceInfo {
```
SHould this be moved into dbs-arch?

jingshan-lx replied:
done

jingshan-lx committed:
dbs-arch: add device information for booting

This modification mainly adds device information for booting, which
mainly involves 3 elements:

1.DeviceType struct: Describes the device type for generating fdt.

2. DeviceInfoForFDT trait: Records device properties for generating fdt.

3.MMIODeviceInfo struct: Record MMIO device information for generating
fdt.

Signed-off-by: jingshan.lx <jingshan@linux.alibaba.com>

jiangliu commented on crates/dbs-arch/src/aarch64/mod.rs:
```diff
@@ -7,3 +7,58 @@
+    }
+
+    fn irq(&self) -> u32 {
+        self.irqs[0]
```
Is there any mechanism to ensure that `self.irqs[0]` is valid?
How should `MMIODeviceInfo::irqs` be organized?

jingshan-lx replied:
Currently only the business process guarantees that self.irqs[0] is valid.
1. Currently MMIODeviceInfo is only used for serial and virtio_mmio devices, which have one and only one irq.
2. MMIODeviceInfo will be created in the create_devices process, and the irqs[0] used in MMIODeviceInfo is in the subsequent system startup process, so the process ensures that self.irqs[0] is valid.
Based on the constraints of the use process in the above two points, is it necessary to add additional judgment conditions to ensure the validity of irqs[0]?

jiangliu replied:
It would be better to return an error instead of panic.
Oh, the function signature doesn't return error code. So maybe an assert to indicate the assumption?

jingshan-lx replied:
done, the return value of the irq() function has been modified to the result type.

jiangliu commented on crates/dbs-arch/src/lib.rs:
```diff
@@ -22,3 +22,24 @@ pub use x86_64::*;
+/// Types of devices that can get attached to this platform.
+#[derive(Clone, Debug, PartialEq, Eq, Hash, Copy)]
+pub enum DeviceType {
+    /// Device Type: Virtio.
```
Any consideration about `PCI`?
PCI is actually supported, and will this structure be used for x86?

jingshan-lx replied:
The specific information of the pci device will not be passed to the guest kernel through fdt. We pass the pci root bus information to fdt, and then the guest kernel confirms the specific pci device information through the probe pci root bus, so there is no pci device in DeviceType.
Re-confirmed the code. Currently, only DeviceType is used in aarch64. Therefore, this structure is placed in the aarch64 directory, and the submission has been re-modified.

jiangliu commented:
Is `output.dtb` used for test? If so, should we move it to `crates/dbs-boot/test`?

jingshan-lx commented:
> 
done

jingshan-lx commented:
> Is `output.dtb` used for test? If so, should we move it to `crates/dbs-boot/test`?

done, I have move it to crates/dbs-boot/src/aarch64/test
